File: modules/vescape-core/ios/sync/SyncJson.swift
import Foundation

/// A row the server could never store. Permanent for this phone: retrying the same bytes cannot make
/// it succeed, so the engine pauses with the row retained rather than skipping it.
///
/// It names the table and the field only — never the value, which may be a coordinate, a Rider's
/// text or a token.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sync/SyncJson.kt `SyncProtocolException`
struct SyncProtocolError: Error, Equatable {
  let table: SyncTable
  let field: String
  let problem: String
}

/// A compact JSON object writer that validates as it writes.
///
/// Deliberately not `JSONSerialization`: this has to produce the exact bytes measured against the
/// wire byte cap, in a stable field order. The bounds it enforces are the server's own
/// (`vescape-server` `src/sync/protocol.ts`), applied before transport so a wedged batch is
/// impossible rather than merely unlikely.
///
/// Nullable columns are written as explicit nulls: "cleared" and "not mentioned" are different
/// intents, and a missing key cannot express the first.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sync/SyncJson.kt `SyncRowWriter`
/// @parity /modules/vescape-server/src/sync/protocol.ts
final class SyncRowWriter {
  private let table: SyncTable
  private var out = "{"

  init(_ table: SyncTable) {
    self.table = table
  }

  func build() -> String { out + "}" }

  /// An identifier the phone chose: a Board id, a settings key, an event name. Never empty.
  @discardableResult
  func keyText(_ field: String, _ value: String) throws -> SyncRowWriter {
    if value.isEmpty { throw fail(field, "must not be empty") }
    return try boundedText(field, value)
  }

  @discardableResult
  func nullableKeyText(_ field: String, _ value: String?) throws -> SyncRowWriter {
    guard let value else { return raw(field, "null") }
    return try keyText(field, value)
  }

  /// A key column the phone derives rather than names, so it may legitimately be empty — a sanitizer
  /// writes `""` as the device id of a sample captured with no Board connected.
  @discardableResult
  func derivedKeyText(_ field: String, _ value: String?) throws -> SyncRowWriter {
    guard let value else { return raw(field, "null") }
    return try boundedText(field, value)
  }

  /// Text the server stores opaquely and hands back unchanged. Uncapped, like the server's.
  @discardableResult
  func text(_ field: String, _ value: String?) -> SyncRowWriter {
    guard let value else { return raw(field, "null") }
    return raw(field, quote(value))
  }

  @discardableResult
  func bool(_ field: String, _ value: Bool) -> SyncRowWriter {
    raw(field, value ? "true" : "false")
  }

  /// Epoch ms, or a duration in ms: non-negative and inside the JSON-safe integer range.
  @discardableResult
  func timestamp(_ field: String, _ value: Int64?) throws -> SyncRowWriter {
    try bounded(field, value, 0, syncSafeIntMax)
  }

  @discardableResult
  func int32(_ field: String, _ value: Int64?) throws -> SyncRowWriter {
    try bounded(field, value, syncInt32Min, syncInt32Max)
  }

  @discardableResult
  func count(_ field: String, _ value: Int64?) throws -> SyncRowWriter {
    try bounded(field, value, 0, syncInt32Max)
  }

  /// A 64-bit column that is not a timestamp — an odometer reading.
  @discardableResult
  func int64(_ field: String, _ value: Int64?) throws -> SyncRowWriter {
    try bounded(field, value, -syncSafeIntMax, syncSafeIntMax)
  }

  /// A real number. Neither infinity nor NaN is expressible in JSON.
  @discardableResult
  func number(_ field: String, _ value: Double?) throws -> SyncRowWriter {
    guard let value else { return raw(field, "null") }
    if !value.isFinite { throw fail(field, "must be finite") }
    let whole = Int64(exactly: value.rounded(.towardZero)) ?? 0
    return raw(field, value == Double(whole) ? String(whole) : String(value))
  }

  private func bounded(_ field: String, _ value: Int64?, _ min: Int64, _ max: Int64) throws -> SyncRowWriter {
    guard let value else { return raw(field, "null") }
    if value < min || value > max { throw fail(field, "is out of bounds") }
    return raw(field, String(value))
  }

  private func boundedText(_ field: String, _ value: String) throws -> SyncRowWriter {
    // UTF-16 code units, matching the server's compiled `value.length <= 128` and Kotlin's
    // `String.length`. Swift's `count` is grapheme clusters, which would let a key through here that
    // the server refuses — and a refused batch is a permanent pause.
    if value.utf16.count > maxSyncKeyLength {
      throw fail(field, "exceeds \(maxSyncKeyLength) characters")
    }
    return raw(field, quote(value))
  }

  @discardableResult
  private func raw(_ field: String, _ encoded: String) -> SyncRowWriter {
    if out.count > 1 { out += "," }
    out += quote(field) + ":" + encoded
    return self
  }

  private func fail(_ field: String, _ problem: String) -> SyncProtocolError {
    SyncProtocolError(table: table, field: field, problem: problem)
  }

  private func quote(_ value: String) -> String {
    var quoted = "\""
    for character in value.unicodeScalars {
      switch character {
      case "\"": quoted += "\\\""
      case "\\": quoted += "\\\\"
      case "\n": quoted += "\\n"
      case "\r": quoted += "\\r"
      case "\t": quoted += "\\t"
      default:
        if character.value < 0x20 {
          quoted += String(format: "\\u%04x", character.value)
        } else {
          quoted.unicodeScalars.append(character)
        }
      }
    }
    return quoted + "\""
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sync/SyncJson.kt
package expo.modules.vescapecore.sync

/**
 * A row the server could never store. Permanent for this phone: retrying the same bytes cannot make
 * it succeed, so the engine pauses with the row retained rather than skipping it.
 *
 * The message names the table and the field only — never the value, which may be a coordinate, a
 * Rider's text or a token.
 *
 * @parity /modules/vescape-core/ios/sync/SyncJson.swift `SyncProtocolError`
 */
class SyncProtocolException(val table: SyncTable, val field: String, val problem: String) :
  IllegalStateException("${table.wire}.$field $problem")

/**
 * A compact JSON object writer that validates as it writes.
 *
 * Deliberately not `org.json`: this has to produce the exact bytes measured against the wire byte
 * cap, in a stable field order, and run in plain JVM tests where the platform's JSON is a stub. The
 * bounds it enforces are the server's own (`vescape-server` `src/sync/protocol.ts`), applied before
 * transport so a wedged batch is impossible rather than merely unlikely.
 *
 * Nullable columns are written as explicit nulls: "cleared" and "not mentioned" are different
 * intents, and a missing key cannot express the first.
 *
 * @parity /modules/vescape-core/ios/sync/SyncJson.swift `SyncRowWriter`
 * @parity /modules/vescape-server/src/sync/protocol.ts
 */
class SyncRowWriter(private val table: SyncTable) {
  private val out = StringBuilder("{")

  fun build(): String = out.append('}').toString()

  /** An identifier the phone chose: a Board id, a settings key, an event name. Never empty. */
  fun keyText(field: String, value: String): SyncRowWriter = apply {
    if (value.isEmpty()) fail(field, "must not be empty")
    boundedText(field, value)
  }

  fun nullableKeyText(field: String, value: String?): SyncRowWriter = apply {
    if (value == null) raw(field, "null") else keyText(field, value)
  }

  /**
   * A key column the phone derives rather than names, so it may legitimately be empty — a sanitizer
   * writes `""` as the device id of a sample captured with no Board connected.
   */
  fun derivedKeyText(field: String, value: String?): SyncRowWriter = apply {
    if (value == null) raw(field, "null") else boundedText(field, value)
  }

  /** Text the server stores opaquely and hands back unchanged. Uncapped, like the server's. */
  fun text(field: String, value: String?): SyncRowWriter = apply {
    if (value == null) raw(field, "null") else raw(field, quote(value))
  }

  fun bool(field: String, value: Boolean): SyncRowWriter = raw(field, if (value) "true" else "false")

  /** Epoch ms, or a duration in ms: non-negative and inside the JSON-safe integer range. */
  fun timestamp(field: String, value: Long?): SyncRowWriter = bounded(field, value, 0, SYNC_SAFE_INT_MAX)

  fun int32(field: String, value: Int?): SyncRowWriter =
    bounded(field, value?.toLong(), SYNC_INT32_MIN, SYNC_INT32_MAX)

  fun count(field: String, value: Int?): SyncRowWriter =
    bounded(field, value?.toLong(), 0, SYNC_INT32_MAX)

  /** A 64-bit column that is not a timestamp — an odometer reading. */
  fun int64(field: String, value: Long?): SyncRowWriter =
    bounded(field, value, -SYNC_SAFE_INT_MAX, SYNC_SAFE_INT_MAX)

  /** A real number. Neither infinity nor NaN is expressible in JSON. */
  fun number(field: String, value: Double?): SyncRowWriter = apply {
    if (value == null) {
      raw(field, "null")
      return@apply
    }
    if (!value.isFinite()) fail(field, "must be finite")
    val whole = value.toLong()
    raw(field, if (value == whole.toDouble()) whole.toString() else value.toString())
  }

  /**
   * A measurement the firmware reported, where non-finite means the reading is unusable rather
   * than that the row is malformed.
   *
   * [number] refuses infinity and NaN, which is right for a value a Rider authored — one there is
   * a bug worth stopping for. A decoded Board sample is the opposite case: the app did not choose
   * the value, it received it, and one bad float would pause every table's backup on a permanent
   * protocol error that no retry can clear. These columns are all nullable precisely because a
   * field the firmware did not send is absent, so an unusable one is absent too.
   *
   * @parity /modules/vescape-core/ios/sync/SyncJson.swift `reading`
   */
  fun reading(field: String, value: Double?): SyncRowWriter =
    number(field, if (value != null && value.isFinite()) value else null)

  private fun bounded(field: String, value: Long?, min: Long, max: Long): SyncRowWriter = apply {
    if (value == null) {
      raw(field, "null")
      return@apply
    }
    if (value < min || value > max) fail(field, "is out of bounds")
    raw(field, value.toString())
  }

  private fun boundedText(field: String, value: String) {
    if (value.length > MAX_SYNC_KEY_LENGTH) fail(field, "exceeds $MAX_SYNC_KEY_LENGTH characters")
    raw(field, quote(value))
  }

  private fun raw(field: String, encoded: String): SyncRowWriter = apply {
    if (out.length > 1) out.append(',')
    out.append(quote(field)).append(':').append(encoded)
  }

  private fun fail(field: String, problem: String): Nothing =
    throw SyncProtocolException(table, field, problem)

  private fun quote(value: String): String {
    val quoted = StringBuilder(value.length + 2).append('"')
    for (char in value) {
      when {
        char == '"' -> quoted.append("\\\"")
        char == '\\' -> quoted.append("\\\\")
        char == '\n' -> quoted.append("\\n")
        char == '\r' -> quoted.append("\\r")
        char == '\t' -> quoted.append("\\t")
        char < ' ' -> quoted.append("\\u%04x".format(char.code))
        else -> quoted.append(char)
      }
    }
    return quoted.append('"').toString()
  }
}
